File: Problem3.py
#Time Complexity :O(n)
#Space Complexity :O(1)
#Did this code successfully run on Leetcode : Yes
#Any problem you faced while coding this : No

class Solution:
    def trap(self, height: List[int]) -> int:
        max_height=max(height)
        max_height_id=height.index(max_height)
        lw=0
        rw=0
        l=0
        r=len(height)-1
        result=0
        while(l<max_height_id):
            if lw>height[l]:
                result+=lw-height[l]
            else:
                lw=height[l]
            l+=1
        while(r>max_height_id):
            if rw>height[r]:
                result+=rw-height[r]
            else:
                rw=height[r]
            r-=1
        return result

File: problem3.java
//## Problem3  Rotate Array by K Places(https://leetcode.com/problems/rotate-array/)

// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach
// we reverse the whole array, then reverse from trarting to k-1, then k to n-1. This replicates the behavior of moving array by k places

class Solution {
    public void rotate(int[] nums, int k) {
        int n= nums.length;
        if(nums==null || nums.length==0 || k%n==0)
        {
            return;
        }

        k=k%n;
        reverse(nums,0,n-1);
        reverse(nums,0,k-1);
        reverse(nums,k,n-1);
    }

    public void reverse(int[] nums, int l, int r)
    {
        int temp;
        while(l<r)
        {
            temp=nums[l];
            nums[l]=nums[r];
            nums[r]=temp;
            l++;
            r--;
        }
    }
}